### DISCUSSION TITLE
How to add controls next to Window system buttons

### DISCUSSION THREAD

**Discussion by ashotjanibekyan** [AvaloniaUI/Avalonia#11818] [Q&A]
Hi. Suppose I would like to add a control element next to the maximize button, like Firefox or Rider

![image](https://github.com/AvaloniaUI/Avalonia/assets/31385299/c3706b78-bda9-4cbe-a7be-e34ab9fd73c5)

![image](https://github.com/AvaloniaUI/Avalonia/assets/31385299/96125131-3587-43cd-93f9-5e7002454bc2)

I tried setting `ExtendClientAreaToDecorationsHint="True"` and using a grid row as a title bar but my controls are under the close/minimize/maximize buttons

![image](https://github.com/AvaloniaUI/Avalonia/assets/31385299/08c2fb26-5c4a-4668-9408-0048ffcf41d5)

Of course, setting the left margin solves the problem but how can I find out the right margin size programmatically?

The same issue exists for Windows version when I'm trying to put something on the left side of Windows buttons.

**Comment by robloo** [comment]
Do the system window buttons actually change size? I would just set the Margin and move on (pretty sure this is what everyone else does).

Otherwise you might be able to find the CaptionButtons in the visual tree and measure width: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/Chrome/CaptionButtons.cs.

**Reply by ashotjanibekyan** [reply]
I'm not sure if they keep the size or not. If they are the same regardless of OS version or screen size/resolution, setting a fixed margin seems the easiest choice.

**Reply by robloo** [reply]
You will have to switch margin based on OS (or choose the max). That seems expected when you are extending into the titlebar which is typically managed by the OS though.

**Reply by ashotjanibekyan** [reply]
By "regardless of OS" I meant Mac versions, of course the margin will be different on Mac and Windows. Sorry for confusion.

**Reply by timunie** [reply]
well, you never know. And we cannot do much about it. Take Windows 11, where suddenly the start button moved into the middle. Who guarntees us, tha the close button will stay where it is 🤷‍♂️? Also if I look back into Windows 95 until today, I think the sizes of the buttons changed between the different versions, so not a constant value can be assumes. 

We, and most likely also you, cannot predict what Windows 12 will do. About the Mac universe, I cannot say much. 

Even worse: In Windows the user can change it, see: https://www.tenforums.com/tutorials/79900-change-size-caption-buttons-windows-10-a.html

**Comment by ashotjanibekyan** [comment]
```xaml
<Style Selector="TitleBar">
    <Setter Property="Template">
        <ControlTemplate>
            <Panel HorizontalAlignment="Right" 
                   VerticalAlignment="Stretch">
                <Panel x:Name="PART_MouseTracker" 
                       Height="1" 
                       VerticalAlignment="Top" />
                <Panel x:Name="PART_Container">
                    <StackPanel Orientation="Horizontal">
                        <TextBlock Text="This is in the title bar"></TextBlock>
                        <Border x:Name="PART_Background"
                                Background="{TemplateBinding Background}"
                                IsHitTestVisible="False"/>
                        <CaptionButtons x:Name="PART_CaptionButtons"
                                        VerticalAlignment="Top" 
                                        HorizontalAlignment="Right"
                                        Foreground="{TemplateBinding Foreground}" />
                    </StackPanel>
                </Panel>
            </Panel>
        </ControlTemplate>
    </Setter>
</Style>
```

this works on Windows but has zero effect on OSX.

**Comment by aneeskA** [comment]
@ashotjanibekyan Were you able to solve this problem?
